### PR TITLE
Update correlation in Apply node

### PR DESCRIPTION
**Pull request by kokosing**
Update correlation in Apply node
    
    All the column references of subquery plan which are not resolved (do
    not come from subquery relations) are collected. SubqueryPlanner check
    if it is able to rewrite such references to symbol produced by its
    relation, if it is able then add replace column reference with that
    symbol and that this symbol to the correlation list of Apply node. If
    SubqueryPlanner is not able to rewrite that column reference, then it is
    assumed the outer SubqueryPlanner execution will do that. If there is no
    outer SubqueryPlanner then SemanticException should be thrown.

### CODE REVIEW

**Comment by kokosing** [comment]
Supersedes https://github.com/prestodb/presto/pull/5302. 
Please notice that this pull request is based on https://github.com/prestodb/presto/pull/5686

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlan.java [hunk 1]
```diff
@@ -58,13 +57,12 @@ public RelationPlan(PlanNode root, Scope scope, List<Symbol> outputSymbols, Opti
     public Optional<Symbol> getSymbol(Expression expression)
     {
         Optional<ResolvedField> resolvedField = scope.tryResolveField(expression);
-        if (!resolvedField.isPresent()) {
-            return Optional.empty();
-        }
 
-        checkState(resolvedField.isPresent() && resolvedField.get().isLocal(), "Unable to get symbol for field in outer scope: '%s'", expression);
+        if (resolvedField.isPresent() && resolvedField.get().isLocal()) {
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlan.java]: `return resolvedField.filter(ResolvedField::isLocal).map(field -> outputSymbols.get(field.getFieldIndex()))`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java [hunk 2]
```diff
@@ -130,7 +132,11 @@ public void put(Expression expression, Symbol symbol)
         expressionToSymbols.put(translated, symbol);
 
         // also update the field mappings if this expression is a field reference
-        analysis.getFieldIndex(expression).ifPresent(index -> fieldSymbols[index] = symbol);
+        Optional<ResolvedField> resolvedField = rewriteBase.getScope().tryResolveField(expression);
+        if (resolvedField.isPresent() && resolvedField.get().isLocal()) {
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java]: use Optional#filter chained with Optional#ifPresent

File: presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java [hunk 3]
```diff
@@ -189,21 +195,27 @@ public Expression rewriteQualifiedNameReference(QualifiedNameReference node, Voi
...
-                Symbol symbol = rewriteBase.getSymbol(fieldIndex.get());
-                checkState(symbol != null, "No symbol mapping for node '%s' (%s)", node, fieldIndex.get());
-                Expression rewrittenExpression = symbol.toSymbolReference();
-
-                return coerceIfNecessary(node, rewrittenExpression);
+                Optional<Symbol> symbol = rewriteBase.getSymbol(node);
+                if (symbol.isPresent()) {
+                    return coerceIfNecessary(node, symbol.get().toSymbolReference());
+                }
+                return node;
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java]:
```
return symbol
    .map(s -> coerceIfNecessary(node, symbol.toSymbolReference()))
    .getOrElse(node);

File: presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java [hunk 4]
```diff
@@ -189,21 +195,27 @@ public Expression rewriteQualifiedNameReference(QualifiedNameReference node, Voi
...
             }
 
             @Override
             public Expression rewriteDereferenceExpression(DereferenceExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
             {
-                if (analysis.getFieldIndex(node).isPresent()) {
-                    return rewriteExpressionWithResolvedName(node);
+                Optional<ResolvedField> resolvedField = rewriteBase.getScope().tryResolveField(node);
+                if (resolvedField.isPresent()) {
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java]: `resolvedField.ifPresent`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java [hunk 5]
```diff
@@ -189,21 +195,27 @@ public Expression rewriteQualifiedNameReference(QualifiedNameReference node, Voi
...
+                    if (resolvedField.get().isLocal()) {
+                        Optional<Symbol> symbol = rewriteBase.getSymbol(node);
+                        checkState(symbol.isPresent(), "No symbol mapping for node '%s'", node);
+                        Expression rewrittenExpression = symbol.get().toSymbolReference();
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java]: This looks kinda similar to what's happening above in rewriteExpressionWithResolvedName. So I'd apply the optional-related comments here as well or maybe even extract a method

Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java]: Also, checking if a reference is local happens suspiciously often - maybe there should be a `resolveLocalField` method?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlan.java (hunk 1 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlan.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java (hunk 2 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java]: done

**Comment by haozhun** [comment]
I did not read this pull request. But I would like to add a comment because of a production issue we've seen before. When Teradata brought non-equi conjunct support to outer join in Presto (thank you again), we removed the sanity check that fails any query that contains such conjuncts (and its tests) before the feature were linked together to execute such query correctly. As a result, for more than a month, the Presto release version produces incorrect result for such queries. 

Please make sure any query with correlated subqueries continues to fail loudly until such query produces correct result.

**Comment by kokosing** [comment]
@haozhun That is a general true. None of the features should be enabled until it works correctly. I am trying to test it as mush as I can, nevertheless the bugs may still exists. That is why code review is so much important, that I could benefit from your experience and possibly broaden the test coverage to avoid that situation you (FB) experienced before.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java (hunk 3 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java (hunk 4 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java]: Sorry, I cannot the `return` is used inside

File: presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java (hunk 5 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/TranslationMap.java]: Done and I tried to apply the second comment but I went more complicated than it was before.

**Comment by kokosing** [comment]
@haozhun Maybe we could display a notice to the user about possible instability of new features, that given feature was recently added and may still have bugs or instabilities. Such notice could be displayed through couple of Presto releases. What do you think about such approach, would it help if you had something like that for non-equi conjunct support to outer join back then?

**Comment by kokosing** [comment]
Addressed comments and rebased.

**Comment by kokosing** [comment]
@martint Rebased. Can you please review?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 6]
```diff
@@ -223,18 +240,137 @@ private PlanBuilder appendSubqueryApplyNode(PlanBuilder subPlan, Expression subq
...
+        return correlation.build();
+    }
+
     private RelationPlan createRelationPlan(Query subquery)
     {
         return new RelationPlanner(analysis, symbolAllocator, idAllocator, metadata, session)
                 .process(subquery, null);
     }
+
+    private Set<Expression> extractMissingExpressions(PlanNode planNode)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: Why is this called extractMissingExpressions? Unless I'm reading this incorrectly, it find all column reference expressions.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 7]
```diff
@@ -223,18 +240,137 @@ private PlanBuilder appendSubqueryApplyNode(PlanBuilder subPlan, Expression subq
...
     }
 
+    private Map<Expression, Symbol> extractCorrelation(PlanBuilder subPlan, PlanNode subquery)
+    {
+        Set<Expression> missingReferences = extractMissingExpressions(subquery);
+        ImmutableMap.Builder<Expression, Symbol> correlation = ImmutableMap.builder();
+        if (!missingReferences.isEmpty()) {
+            for (Expression missingReference : missingReferences) {
+                Expression rewritten = subPlan.rewrite(missingReference);
+                if (rewritten != missingReference) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: Is reference equality intentional? It seem fragile to be relying on anything like that here.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 8]
```diff
@@ -223,18 +240,137 @@ private PlanBuilder appendSubqueryApplyNode(PlanBuilder subPlan, Expression subq
...
+                            ImmutableList.copyOf(correlation.values())),
                     subPlan.getSampleWeight());
         }
     }
 
+    private Map<Expression, Symbol> extractCorrelation(PlanBuilder subPlan, PlanNode subquery)
+    {
+        Set<Expression> missingReferences = extractMissingExpressions(subquery);
+        ImmutableMap.Builder<Expression, Symbol> correlation = ImmutableMap.builder();
+        if (!missingReferences.isEmpty()) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: This is unnecessary. The loop below will short-circuit if the set is empty.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 9]
```diff
@@ -223,18 +240,137 @@ private PlanBuilder appendSubqueryApplyNode(PlanBuilder subPlan, Expression subq
...
+            for (Expression missingReference : missingReferences) {
+                Expression rewritten = subPlan.rewrite(missingReference);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: This logic needs some comments. It not obvious what it's trying to do.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java [hunk 10]
```diff
@@ -223,18 +240,137 @@ private PlanBuilder appendSubqueryApplyNode(PlanBuilder subPlan, Expression subq
...
+    {
+        private final PlanNodeIdAllocator idAllocator;
+        private final Map<Expression, Expression> mapping;
+
+        public ExpressionReplacer(PlanNodeIdAllocator idAllocator, Map<Expression, Expression> mapping)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.mapping = requireNonNull(mapping, "mapping is null");
+        }
+
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: What's the expected behavior for unsupported plan nodes?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java [hunk 11]
```diff
@@ -526,35 +527,40 @@ private Void visitSetOperation(SetOperationNode node, Void context)
...
+                    .addAll(correlation)
+                    .addAll(node.getCorrelation())
+                    .build();
+
+            node.getInput().accept(this, correlation); // visit child
+            node.getSubquery().accept(this, subqueryCorrelation); // visit child
 
             checkDependencies(node.getInput().getOutputSymbols(), node.getCorrelation(), "APPLY input must provide all the necessary correlation symbols for subquery");
-            checkDependencies(DependencyExtractor.extractUnique(node.getSubquery()), node.getCorrelation(), "not all APPLY correlation symbols are not used in subquery");
+//            checkDependencies(DependencyExtractor.extractUnique(node.getSubquery()), node.getCorrelation(), "not all APPLY correlation symbols are used in subquery");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: remove commented-out code

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java [hunk 12]
```diff
@@ -568,6 +574,14 @@ private void verifyUniqueId(PlanNode node)
 
             nodesById.put(id, node);
         }
+
+        private ImmutableSet<Symbol> createInputs(PlanNode source, Collection<Symbol> correlation)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: use `Set<Symbol>` in the method signature

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java [hunk 13]
```diff
@@ -79,40 +80,40 @@
...
 
     private static class Visitor
-            extends PlanVisitor<Void, Void>
+            extends PlanVisitor<List<Symbol>, Void>
     {
         private final Map<PlanNodeId, PlanNode> nodesById = new HashMap<>();
 
         @Override
-        protected Void visitPlan(PlanNode node, Void context)
+        protected Void visitPlan(PlanNode node, List<Symbol> correlation)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: I'd change the term `correlations` to `boundSymbols` or `boundVariables` (as in "free" vs "bound" variables in programming languages)

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java [hunk 14]
```diff
@@ -5880,53 +5880,60 @@ public void testExistsSubqueryWithGroupBy()
...
-        assertQueryFails("SELECT * FROM lineitem l ORDER BY (SELECT l.orderkey)", errorMsg);
+        assertQueryFails("SELECT (SELECT 2 * l.orderkey) FROM lineitem l", errorMsg);
+        assertQueryFails("SELECT * FROM lineitem l WHERE 1 = (SELECT 2 * l.orderkey)", errorMsg);
+        assertQueryFails("SELECT * FROM lineitem l ORDER BY (SELECT 2 * l.orderkey)", errorMsg);
 
         // group by
-        assertQueryFails("SELECT max(l.quantity), l.orderkey, (SELECT l.orderkey) FROM lineitem l GROUP BY l.orderkey", errorMsg);
-        assertQueryFails("SELECT max(l.quantity), l.orderkey FROM lineitem l GROUP BY l.orderkey HAVING max(l.quantity) < (SELECT l.orderkey)", errorMsg);
-        assertQueryFails("SELECT max(l.quantity), l.orderkey FROM lineitem l GROUP BY l.orderkey, (SELECT l.orderkey)", errorMsg);
+        assertQueryFails("SELECT max(l.quantity), 2 * l.orderkey, (SELECT l.orderkey) FROM lineitem l GROUP BY l.orderkey", errorMsg);
```
Comment (martint) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: Why did these change?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java [hunk 15]
```diff
@@ -701,13 +701,26 @@ public PlanNode visitValues(ValuesNode node, RewriteContext<Set<Symbol>> context
...
                     .addAll(context.get())
-                    .addAll(node.getCorrelation())
+                    .addAll(newCorrelation)
                     .build();
             PlanNode input = context.rewrite(node.getInput(), inputContext);
-            PlanNode subquery = context.rewrite(node.getSubquery(), context.get());
-            return new ApplyNode(node.getId(), input, subquery, node.getCorrelation());
+
+            if (!inputContext.stream().anyMatch(subquery.getOutputSymbols()::contains)) {
+                // prune ApplyNode as subquery symbols are not used
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java]: This is only valid if the subquery is guaranteed to produce at most 1 row, no?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java [hunk 16]
```diff
@@ -70,6 +70,11 @@
 
 public class LogicalPlanner
 {
+    public enum Stage
+    {
+        CREATED, OPTIMIZED, OPTIMIZED_AND_VALIDATED
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: This should probably in a separate commit.

Also, can you remind me of the rationale for this? I remember discussing about this, but I can't find the original comments.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java [hunk 17]
```diff
@@ -866,7 +867,16 @@ public Void visitEnforceSingleRow(EnforceSingleRowNode node, Integer indent)
         }
 
         @Override
-        protected Void visitPlan(PlanNode node, Integer context)
+        public Void visitApply(ApplyNode node, Integer indent)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java]: This should go in a separate commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java [hunk 18]
```diff
@@ -180,7 +180,7 @@ public static Object evaluateConstantExpression(Expression expression, IdentityH
             @Override
             public Expression rewriteExpression(Expression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
             {
-                Expression rewrittenExpression = treeRewriter.defaultRewrite(node, context);
+                Expression rewrittenExpression = treeRewriter.defaultRewrite(node, null);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Why did these change?

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionAliases.java [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+final class ExpressionAliases
+{
+    private final Multimap<String, Expression> map;
+
+    ExpressionAliases()
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionAliases.java]: I'd make this public

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionAliases.java [hunk 20]
```diff
@@ -0,0 +1,71 @@
...
+    {
+        this.map = ArrayListMultimap.create();
+    }
+
+    ExpressionAliases(ExpressionAliases expressionAliases)
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionAliases.java]: Same here

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionVerifier.java [hunk 21]
```diff
@@ -71,6 +70,16 @@ protected Boolean visitNode(Node node, Expression context)
     }
 
     @Override
+    protected Boolean visitInPredicate(InPredicate actual, Expression expectedExpression)
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionVerifier.java]: These new methods should go in a separate commit (vs the one that adds support for matching apply in plan assertion dsl)

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionVerifier.java [hunk 22]
```diff
@@ -83,13 +92,28 @@ protected Boolean visitComparisonExpression(ComparisonExpression actual, Express
...
-            LongLiteral expected = (LongLiteral) expectedExpression;
-            return actual.getValue() == expected.getValue();
+        if (expression instanceof LongLiteral) {
+            return String.valueOf(((LongLiteral) expression).getValue());
+        }
+        else if (expression instanceof GenericLiteral) {
+            return ((GenericLiteral) expression).getValue();
+        }
+        else {
+            throw new IllegalArgumentException("Not supported literal expression");
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionVerifier.java]: "Unsupported literal expression type: " + expression.getClass().getName()

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java [hunk 23]
```diff
@@ -71,7 +75,7 @@ public static PlanMatchPattern anyNot(Class<? extends PlanNode> excludeNodeClass
 
     public static PlanMatchPattern tableScan(String expectedTableName)
     {
-        return any().with(new TableScanMatcher(expectedTableName));
+        return node(TableScanNode.class).with(new TableScanMatcher(expectedTableName));
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: These changes should probably go in a separate commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java [hunk 24]
```diff
@@ -249,20 +249,22 @@ else if (node.getSource().getOutputSymbols().contains(symbol)) {
...
-                return new SemiJoinNode(idAllocator.getNextId(),
-                        node.getInput(),
-                        node.getSubquery(),
-                        value,
-                        valueList,
-                        semiJoinSymbol.get(),
-                        Optional.empty(),
-                        Optional.empty()
-                );
+            if (node.getCorrelation().isEmpty()) {
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java]: If this is a bug, can you add a test that reproduces the problem?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java (hunk 8 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java (hunk 9 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java (hunk 7 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: I changed that to check to `rewritten instanceof SymbolReference`.

Moreover I extracted method and added comment to make it more clear what it is going to do.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java (hunk 6 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: I will rename it. It finds missing column reference expressions i.e. column reference which cannot be satisfied within given PlanNode.

At this point all column reference should be rewritten to symbol references. If it finds column reference that means this is a reference to column from outer scope

File: presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java (hunk 10 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/SubqueryPlanner.java]: IMO all nodes are supported. Default behavior is to visit children nodes (SimplePlanRewriter). 

However, the problem here is that ExpressionReplacer can miss some nodes which requires expression to rewritten (case of some visit* missing). As all such method has to be listed (handled) here explicitly. But this is a common problem with visitor pattern.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java (hunk 13 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java (hunk 11 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: this should not be commented.

Fixed.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java (hunk 12 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/ValidateDependenciesChecker.java]: fixed

File: presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java (hunk 14 above)
Comment (kokosing) [inline presto-tests/src/main/java/com/facebook/presto/tests/AbstractTestQueries.java]: To make sure that column references are extracted from expressions. During the development I had a bug that I treated whole expression as column reference.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java (hunk 15 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java]: No.

This handles queries like: 
```
SELECT COUNT(*) FROM (
  SELECT orderkey IN (SELECT orderkey 
    FROM lineitem 
    WHERE orderkey % 2 = 0)
  FROM orders LIMIT 10)
```

Please notice that actual result of IN predicate is not used and so ApplyNode should be removed regardless how many rows has input relation.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java (hunk 18 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java]: Good question :) Restored.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java (hunk 15 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java]: What about something like:

```sql
SELECT COUNT(*) FROM (
    SELECT * FROM a CROSS JOIN LATERAL (SELECT * FROM b WHERE b.m = a.x)
)
```

We don't support LATERAL yet, but when we do...

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PruneUnreferencedOutputs.java]: Eh... I really don't like when you are right ;)

It is much more complicated than I previously thought. `Apply` node can be only pruned when it is placed in projection (or in `VALUES`) and its output is not used. To do that I need to remove `Apply` when I prune matching `Expression`. In this place I have no context what introduced given `Apply` node so I cannot be removed.

I am going to rework this.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java (hunk 16 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/LogicalPlanner.java]: Fixed. I added a commit message which explains the motivation.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java (hunk 17 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/PlanPrinter.java]: done

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionAliases.java (hunk 19 above)
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionAliases.java]: done

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionAliases.java (hunk 20 above)
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionAliases.java]: done

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionVerifier.java (hunk 22 above)
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionVerifier.java]: done

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionVerifier.java (hunk 21 above)
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/ExpressionVerifier.java]: done

File: presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java (hunk 23 above)
Comment (kokosing) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java]: done

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java (hunk 24 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java]: So far when there is correlation `StatementAnalyzer` raises exception, so I didn't have to check if node is correlated. 

One of the previous (in this PR) commit changes that and updates correlation and so this rewrite has to check if apply is correlated.

I moved this commit before the commit which updates correlation.

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/TransformUncorrelatedInPredicateSubqueryToSemiJoin.java]: It is covered by tests for correlation.

**Comment by kokosing** [comment]
@martint Rebased. Comments applied. Superseded by #5943
